init.php

```diff
@@ -1,121 +1,257 @@
+                    trigger_error( $e->getMessage(), E_USER_WARNING );
+                }
+
+                restore_error_handler();
+
+            }
+
+            if ( ! $visible || isset( $hidden_items[ $item_parent ] ) ) { // also hide the children of invisible items
+                unset( $items[ $key ] );
+                $hidden_items[ $item->ID ] = '1';
+            }
+        }
+
+        return $items;
+
+    }
+
+    /**
+     * Remove the _menu_item_visibility meta when the menu item is removed
+     *
+     * @param int $post_id
+     *
+     * @return void
+     *
+     * @since 0.2.2
+     */
+    public static function remove_visibility_meta( int $post_id ) : void {
+        if ( is_nav_menu_item( $post_id ) ) {
+            delete_post_meta( $post_id, '_menu_item_visibility' );
+        }
+    }
+
+    /**
+     * Handle errors in eval-ed Logic field
+     *
+     * @param int $errno
+     * @param string $errstr
+     * @param string $errFile
+     * @param int $errLine
+     * @param array $errContext
+     *
+     * @return void
+     *
+     * @since 0.4
+     */
+    public static function error_handler( int $errno, string $errstr, string $errFile, int $errLine, array $errContext ) : bool {
+
+        if ( current_user_can( 'manage_options' ) ) {
+
+            if ( ! empty( $errContext['item']->title ) ) {
+                echo sprintf( __( 'Error in "%s" menu item Visibility: ', 'menu-items-visibility-control' ), $errContext['item']->title );
+            }
+
+            echo $errstr;
+
+        }
+
+        /* Don't execute PHP internal error handler */
+        return true;
+
+    }
+
+    /**
+     * Check if methods run through eval() are safe.
+     *
+     * @param array $menu_item_visibility
+     *
+     * @return bool
+     *
+     * @since 2021.07.21
+     */
+    public static function eval_security_check( array $menu_item_visibility ) : bool {
+
+        $tokens = implode( ',', $menu_item_visibility );
+        $tokens = '<?php ' . $tokens . ' ?>';
+        $tokens = token_get_all( $tokens );
+
+        $cxl_eval_safe_tokens = get_option( 'cxl_eval_safe_tokens' );
+
+        foreach ( $tokens as $i => $token ) {
+
+            // Identify method via '(' symbol.
+            if ( ! is_array( $token ) && '(' === $token ) {
+
+                // Previous token was a method.
+                if ( ! in_array( $tokens[$i-1][1], $cxl_eval_safe_tokens ) ) {
+                    return false;
+                }
+
+            }
+
+        }
+
+        return true;
+
+    }
+
+
+
+    /**
+     * Register eval() safe tokens.
+     *
+     * @return void
+     *
+     * @since 2021.07.21
+     */
+    public static function register_eval_safe_tokens() : void {
+
+        $cxl_eval_safe_tokens = [
+            'count',
+            'empty',
+            'get_current_user_id',
+            'wc_memberships_get_user_active_memberships',
+            'wc_memberships_is_user_active_member',
+        ];
```

Comment:
@martinbon what conditional functionals do we look to deploy at first?

I'm not sure where @paulkirspuu got this list from.

Reply:
Here's how we display the Foundation plan button on staging:
`wc_memberships_is_user_active_member(get_current_user_id(), 'foundational-plan' ) && ! empty( wc_memberships_get_user_active_memberships( get_current_user_id() ) ) && 1 === count( wc_memberships_get_user_active_memberships( get_current_user_id() ) )`

This was the first goal that we want to achieve.

Reply:
`empty()` should not be used anywhere.

Propose something that avoids repeating `wc_memberships_get_user_active_memberships( get_current_user_id() )`.

Reply:
Technically, can remove the `! empty()` - Since we can try to do the CTA with "Start free trial for $1" that already exists. I don't have a good answer on how to remove the second `get_current_user_id`